include/sparrow/buffer_view.hpp

```diff
@@ -0,0 +1,270 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    /*
+     * @class buffer_view
+     * @brief Object that references but does not own a piece of contiguous memory
+     */
+    template <class T>
+    class buffer_view
+    {
+    public:
+
+        using self_type = buffer_view<T>;
+        using value_type = T;
+        using reference = T&;
+        using const_reference = const T&;
+        using pointer = T*;
+        using const_pointer = const T*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using iterator = pointer_iterator<pointer>;
+        using const_iterator = pointer_iterator<const_pointer>;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+        explicit buffer_view(buffer<T>& buffer);
+        buffer_view(pointer p, size_type n);
+
+        bool empty() const noexcept;
+        size_type size() const noexcept;
+
+        reference operator[](size_type);
+        const_reference operator[](size_type) const;
+
+        reference front();
+        const_reference front() const;
+
+        reference back();
+        const_reference back() const;
+
+        template <class U = T>
+        U* data() noexcept;
+
+        template <class U = T>
+        const U* data() const noexcept;
+
+        iterator begin();
+        iterator end();
+
+        const_iterator begin() const;
+        const_iterator end() const;
+        const_iterator cbegin() const;
+        const_iterator cend() const;
+
+        reverse_iterator rbegin();
+        reverse_iterator rend();
+
+        const_reverse_iterator rbegin() const;
+        const_reverse_iterator rend() const;
+        const_reverse_iterator crbegin() const;
+        const_reverse_iterator crend() const;
+
+        void swap(buffer_view& rhs) noexcept;
+        bool equal(const buffer_view& rhs) const;
+
+    private:
+
+        pointer p_data = nullptr;
+        size_type m_size = 0u;
+    };
+
+    template <class T>
+    bool operator==(const buffer_view<T>& lhs, const buffer_view<T>& rhs);
```

Comment:
Why not declare it in `buffer_view` as friend?

BTW should we enable comparison with another type of (contiguous?) range?

Reply:
`operator!=` is missing?

Reply:
`operator!=` is now (C++20) automatically generated when you provide `operator==`

Reply:
> `operator!=` is missing?

In C++20, it is automatically defined after `operator==`'s definition.